flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/HMM.scala

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.classification
+
+import org.apache.flink.api.scala.DataSet
+
+import org.apache.flink.api.scala._
+import org.apache.flink.ml.math.{DenseMatrix, DenseVector}
+import org.apache.flink.ml.optimization.BaumWelch
+import org.apache.flink.util.Collector
+import org.slf4j.LoggerFactory
+
+class HMM {
+
+  /**
+   * initial probabilities of the states
+   */
+  var initialProbabilities: DenseVector = null
+
+
+  /**
+   * transition probabilities
+   */
+  var transitionProbabilities: DenseMatrix = null
+
+
+  /**
+   * emission probabilities
+   */
+  var emissionProbabilities: DenseMatrix = null
+
+  val LOG = LoggerFactory.getLogger(this.getClass)
+
+  /**
+   *
+   * Initializes the parameters of the HMM with labeled training data
+   *
+   * @param trainingData labeled training data with the first element of the tuple as
+   *                     the observation and the second element as the label
+   * @param numberOfStates number of states
+   * @param numberOfObVocab number of the observation vocabulary
+   */
+  def initializeParameters(trainingData: DataSet[Array[Tuple2[Int, Int]]],
+                           numberOfStates: Int, numberOfObVocab: Int) = {
+
+    val params =
+
+      trainingData.flatMap((data, collector: Collector[(String, Int, Int, Long)]) => {
+
+      var transitions : Map[String, Long] = Map.empty[String, Long].withDefaultValue(0)
+      var emissions : Map[String, Long] = Map.empty[String, Long].withDefaultValue(0)
+
+      var currentOb = data(0)
+
+      //xi -> yi
+      var emission: String = currentOb._2 + "->" + currentOb._1
+
+      emissions += emission -> 1
+
+      val initialState = ("i", currentOb._2, 0, 1L)
+
+      for (i <- 1 until data.length) {
+        val nextOb= data(i)
+        val transition = currentOb._2 + "->" + nextOb._2
+        currentOb = nextOb
+        emission = currentOb._2 + "->" + currentOb._1
+        transitions += (transition -> (transitions(transition) + 1L))
+        emissions += (emission -> (emissions(emission) + 1L))
+      }
+
+      //emit initial state
+      collector.collect(initialState)
+
+      //emit transitions
+      transitions.foreach(t => {val split = t._1.split("->");
+
+        collector.collect(("t", split(0).toInt,split(1).toInt, t._2))})
+
+        //emit emissions
+       emissions.foreach(e => {val split = e._1.split("->");
+       collector.collect(("e", split(0).toInt, split(1).toInt, e._2))})
+    })
+        //group each tuple types and sum them up
+      .groupBy(0, 1,2).sum(3).map(tuple => (tuple._1, tuple._2,
+        List((tuple._3, tuple._4)), tuple._4))
+        .groupBy(0, 1).reduce((tuple1, tuple2) =>
+        (tuple1._1, tuple1._2, tuple1._3 ::: tuple2._3, tuple1._4 + tuple2._4)).collect()
+
+
+    initialProbabilities = new DenseVector(new Array[Double](numberOfStates))
+    transitionProbabilities = new DenseMatrix(numberOfStates, numberOfStates,
```

Comment:
Even if there are few states, there can be millions or even billions of training sequences which have to be processed in parallel. 
I'm not so familiar with problems which can be solved using HMM. For example, in NLP there are only a few states.